### PR TITLE
Implement HMAC SHA-1 and HMAC-SHA-256 throughout Okio.

### PR DESCRIPTION
**Pull request by swankjesse**
This adds APIs to ByteString, Buffer, HashingSource and HashingSink.

I've forgone polymporphism in HashingSource and HashingSink because it
adds more code than it saves. I can be convinced here.

Closes: https://github.com/square/okio/issues/254

### CODE REVIEW

**Comment by adriancole** [comment]
Thanks tons! I'd use this right away in Zipkin for its amazon support
https://github.com/openzipkin/zipkin/blob/master/zipkin-autoconfigure/storage-elasticsearch-aws/src/main/java/zipkin/autoconfigure/storage/elasticsearch/aws/AWSSignatureVersion4.java#L175

File: okio/src/main/java/okio/HashingSource.java [hunk 1]
```diff
@@ -79,7 +107,11 @@ private HashingSource(Source source, String algorithm) {
       // Hash that segment and all the rest until the end.
       while (offset < sink.size) {
         int pos = (int) (s.pos + start - offset);
-        messageDigest.update(s.data, pos, s.limit - pos);
+        if (messageDigest != null) {
+          messageDigest.update(s.data, pos, s.limit - pos);
+        } else if (mac != null) {
```
Comment (JakeWharton) [inline okio/src/main/java/okio/HashingSource.java]: Any reason you went for `else if` vs. just `else`?

File: okio/src/main/java/okio/Buffer.java [hunk 2]
```diff
@@ -1538,16 +1541,46 @@ public ByteString sha256() {
...
+  public ByteString hmacSha1(ByteString key) {
+    return hmac("HmacSHA1", key);
+  }
+
+  /** Returns the 256-bit SHA-256 HMAC of this buffer. */
+  public ByteString hmacSha256(ByteString key) {
+    return hmac("HmacSHA256", key);
+  }
+
+  private ByteString hmac(String algorithm, ByteString key) {
```
Comment (cketti) [inline okio/src/main/java/okio/Buffer.java]: Missing `null` check for `key`.

File: okio/src/main/java/okio/ByteString.java [hunk 3]
```diff
@@ -150,6 +153,28 @@ private ByteString digest(String algorithm) {
...
+  public ByteString hmacSha1(ByteString key) {
+    return hmac("HmacSHA1", key);
+  }
+
+  /** Returns the 256-bit SHA-256 HMAC of this byte string. */
+  public ByteString hmacSha256(ByteString key) {
+    return hmac("HmacSHA256", key);
+  }
+
+  private ByteString hmac(String algorithm, ByteString key) {
```
Comment (cketti) [inline okio/src/main/java/okio/ByteString.java]: Missing `null` check for `key`.

File: okio/src/main/java/okio/HashingSink.java [hunk 4]
```diff
@@ -55,12 +59,36 @@ public static HashingSink sha256(Sink sink) {
...
     super(sink);
     try {
       this.messageDigest = MessageDigest.getInstance(algorithm);
+      this.mac = null;
+    } catch (NoSuchAlgorithmException e) {
+      throw new AssertionError();
+    }
+  }
+
+  private HashingSink(Sink sink, ByteString key, String algorithm) {
```
Comment (cketti) [inline okio/src/main/java/okio/HashingSink.java]: Missing `null` check for `key`.

File: okio/src/main/java/okio/HashingSource.java [hunk 5]
```diff
@@ -53,12 +57,36 @@ public static HashingSource sha256(Source source) {
...
     super(source);
     try {
       this.messageDigest = MessageDigest.getInstance(algorithm);
+      this.mac = null;
+    } catch (NoSuchAlgorithmException e) {
+      throw new AssertionError();
+    }
+  }
+
+  private HashingSource(Source source, ByteString key, String algorithm) {
```
Comment (cketti) [inline okio/src/main/java/okio/HashingSource.java]: Missing `null` check for `key`.

File: okio/src/main/java/okio/Buffer.java (hunk 2 above)
Comment (cketti) [inline okio/src/main/java/okio/Buffer.java]: Right. I wasn't sure whether explicit checks were omitted intentionally. It's not without precedent in the codebase. But I had the impression explicit checks are the norm.
Is there a code style document where preferences for things like this are spelled out?

Comment (swankjesse) [inline okio/src/main/java/okio/Buffer.java]: We don’t have a document. Closest is the Google Java Style Guide and Effective Java.

File: okio/src/main/java/okio/HashingSource.java (hunk 1 above)
Comment (swankjesse) [inline okio/src/main/java/okio/HashingSource.java]: None. Fixed.

**Pull request merged**